File: 0093_restoreIPAddresses/main.cpp
#include <vector>
#include <string>
#include <algorithm>

class Solution {
private:
    void backtrack(int startIdx, int dots, std::string curIP, std::string& s, std::vector<std::string>& res) {
        if(dots == 4 && startIdx == s.size()) {
            res.push_back(curIP.substr(0, curIP.size()-1));
            return;
        }
        if(dots > 4)
            return;

        for(int i = startIdx; i < std::min(startIdx+3, (int)s.size()); ++i) {
            if(std::stoi(s.substr(startIdx, i - startIdx + 1)) <= 255 && (s[startIdx] != '0' || startIdx == i))
                backtrack(i+1, dots+1, curIP + s.substr(startIdx, i - startIdx + 1) + '.', s, res);
        }
    }
public:
    std::vector<std::string> restoreIpAddresses(std::string s) {
        std::vector<std::string> res;
        if(s.size() <= 12)
            backtrack(0, 0, "", s, res);
        return res;
    }
};

int main() {
    std::string test = "0000";
    Solution sol;
    sol.restoreIpAddresses(test);
    return 0;
}

File: 0093_restoreIPAddresses/main.py
from typing import List
    
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = []
        if len(s) > 12:
            return res
        
        def backtrack(startIdx, dots, curIP):
            if dots == 4 and startIdx == len(s):
                res.append(curIP[:-1])
                return
            if dots > 4:
                return
                
            for i in range(startIdx, min(startIdx + 3, len(s))):
                if int(s[startIdx:i+1]) <= 255 and (s[startIdx] != '0' or startIdx == i):
                    backtrack(i+1, dots + 1, curIP + s[startIdx:i+1] + '.')

        backtrack(0, 0, "")
        return res
            